### ISSUE TITLE
Potential JupyterLab partnership with Tidelift

### ISSUE THREAD

**Issue by andrii-i** [jupyter-governance/ec-team-compass#20]
FYI there is a discussion going in JupyterLab team compass about potential partnership with Tidelift (https://tidelift.com/).  Tidelift offers a partnership that provides monthly income to open source efforts like Project Jupyter. Potential $500-600/month. They do this in exchange for project committing to maintaining good security practices.

Please see discussions thread for details https://github.com/jupyterlab/team-compass/discussions/224.

**Comment by Ruv7** [comment]
Hi Andrii - the link you provide loops back to the EC Team Compass main page. Can you take a look at that and update or explain?

**Comment by krassowski** [comment]
Any feedback from EC on this? It would be amazing to get help with setting this up.

**Comment by Ruv7** [comment]
Thanks for flagging this, we have a substantial backlog right now with the focus on conversations around fiscal sponsorship. 

I'm looking at [224](https://github.com/jupyterlab/frontends-team-compass/discussions/224) and seeing the next steps suggested by Matthias. I want to say the conversation should likely continue there and then you reach out to the EC when you're ready for the green light to deposit into the official bank accounts. Unless there is some other actionable item I would propose closing this issue until further progress has been made.

**Comment by Zsailer** [comment]
Apologies for the slow reply here; we've been working really hard over the last year to get the Jupyter Foundation in place, so this has been lower priority temporarily. I'll work on picking this back up. We've been in communication with folks at Tidelift and LF Charities to get an agreement in place. Once established, we can route Tidelift to Jupyter under LF Charities.

This money is separate from the money collected by the Jupyter Foundation through memberships, so its not controlled by Jupyter Foundation Governing Board. Instead, this money will be managed by the Jupyter EC. It would be great to discuss ways to use this money in a separate thread.

**Comment by krassowski** [comment]
For record, posting my email to EC, dated 21 Aug 2024, in reply to Ana's email after discussing this during EC office hours.

---

Hi Ana & EC,

> I'm gathering info / asking questions from Linux Foundation folks to find out what their experience (if any) has been with this program

Amazing, thank you for that!

> When I visit the website and search for Jupyter the list below is what I see.

Thanks for sharing the screenshot - when I search "jupyter" I see a different list, and in fact I see a different list every other time I search so I think tidelift search does not work very well.

In addition to packages published on PyPI, Jupyter Frontends also publish packages on NPM.
Back when Matthias posted https://github.com/jupyterlab/frontends-team-compass/discussions/224 last year each npm package was generating 25$ but it no longer shows up as of today, it might be because there were no "lifters" and sponsors decided it is not worth paying tidelift anymore, or maybe tidelift is just not showing smaller sums anymore.
To find the npm packages which are under control of Jupyter Project you can use these links:
- https://tidelift.com/lifter/search?q=npm:jupyterlab-release-bot - 179 packages
- https://tidelift.com/lifter/search?q=npm:jupyter-release-bot - 32 packages

I would say we could enroll all of them, but the effort may be prohibitive. I did a quick check across sub projects and it looks like indeed for Frontends the projects with current estimates are generating 600-700$ - each of the projects listed below is given an estimate of 100$:

Frontends subproject:

1. notebook: https://tidelift.com/lifter/search/pypi/notebook
2. jupyterlab: https://tidelift.com/lifter/search/pypi/jupyterlab
3. jupyterlab-server: https://tidelift.com/lifter/search/pypi/jupyterlab-server
4. nbclassic: https://tidelift.com/lifter/search/pypi/nbclassic
5. jupyterlab-pygments: https://tidelift.com/lifter/search/pypi/jupyterlab-pygments
6. notebook-shim: https://tidelift.com/lifter/search/pypi/notebook-shim
7. jupyter-lsp*: https://tidelift.com/lifter/search/pypi/jupyter-lsp (*still has an ambiguous status, but personally I would say enroll and funnel into Project Jupyter account)

Server subproject

1. jupyter-server: https://tidelift.com/lifter/search/pypi/jupyter-server
2. jupyter-server-terminals: https://tidelift.com/lifter/search/pypi/jupyter-server-terminals
3. terminado: https://tidelift.com/lifter/search/pypi/terminado
4. jupyter-events: https://tidelift.com/lifter/search/pypi/jupyter-events

Kernels subproject

1. ipython (already claimed): https://tidelift.com/lifter/search/pypi/ipython
2. ipykernel: https://tidelift.com/lifter/search/pypi/ipykernel

Widgets subproject

1. ipywidgets: https://tidelift.com/lifter/search/pypi/ipywidgets
2. jupyterlab-widgets: https://tidelift.com/lifter/search/pypi/jupyterlab-widgets

Foundations and Standards

1. nbformat: https://tidelift.com/lifter/search/pypi/nbformat
2. nbconvert: https://tidelift.com/lifter/search/pypi/nbconvert
3. nbclient: https://tidelift.com/lifter/search/pypi/nbclient
4. jupyter-client: https://tidelift.com/lifter/search/pypi/jupyter-client
5. jupyter-core: https://tidelift.com/lifter/search/pypi/jupyter-core

Note: while nbdime, nbgrader, and nbviewer do not show an estimate right now, if you were to enroll other Jupyter projects I would enroll these now as they do not show up in simple "jupyter" search.
I did not look at jupyterhub packages, but I know that folks in JupyterHub were doing an amazing job patching vulnerabilities recently - maybe there are lessons to be learned from them.

So that is in total ~2000$ a month that could be earmarked for security across Jupyter (assuming other subprojects are fine with enrolling). If the npm packages generate even 5$ a month each, this might also increase the number by another 1000$ but I do not know if this is the case right now.

> I want to be sure we understand your vision for how this could work

I do not have a strictly formulated vision. I think it would be nice if we could make the funds available:

- to cover expenses when individual subprojects triage/patch security vulnerabilities
- for jupyter security suproject to do whatever they find beneficial with the remainder

I tried to formulate some thoughts on how covering expenses should work:

- objectives - it feels like there should be:
  - a) incentives for folks with different affiliations to pick up the work, especially if they have relevant expertise to solve the problem cleanly
  - b) some kind of assurance that funds are not spent excessively on low-priority work like addressing false positive alerts from npm ecosystem - so that it is available when actually needed for (a)
  - c) some thought put to not create a bad incentive structure like "I do not need to worry about security of my PR because later I will get paid to fix vulnerabilities anyways"
- deciding on spending:
  - maybe the model of "everyone can claim, we are all adults and everyone will behave nicely" could work - I don't know
  - maybe it could be a (part-time?) core maintainer/coordinator employed by LF/Jupyter who would reach out to different devs (based on relevant experience and availability) to find someone to work on a vulnerability patch
     - having a single person do this may be a single point of failure
  - maybe the model of "let's vote in a PR if we need to" as in jupyterhub (https://jupyterhub-team-compass.readthedocs.io/en/latest/practices/funding.html) could/could not work; the challenges here are:
    - we would need to reveal the security vulnerability to all subproject members (even if vote takes place in a private repo, they would need to know what they are voting on);
        - I think this may be a ok because we can just state assessed impact, and all members with access to it can contest the impact
     - in the time it takes to vote asynchronously, the vulnerability could be exploited; while I am sure folks would step in and give up their own time anyways (as it is the case currently), this does not work for incentivizing a larger set of contributors to tackle these issues
      - sometimes vulnerabilities affect multiple subprojects, then it would not be clear who is to vote on fund allocation
  - In any case we should codify this in a governance document after broader consultation with the Security Subproject and SSC. Some inspiration can be taken from https://numpy.org/neps/nep-0048-spending-project-funds.html which explicitly mentions Tidelift.
  - there is a question of how the actual payouts/compensation would work [...] I think it is of the prime importance to find a model flexible enough so that it could encourage experienced folks from as many different companies to take part as possible

While this discussion started around the Jupyter Frontends subproject, it might be worth considering a Jupyter-wide solution. I know that Matthias also has some thoughts, but I will let him voice them once we move to a public discussion medium. One point that I will relay is that we can start with claiming the funds and deciding on the exact processes later (in public).

I'm curious what you all think about the above and excited to hear other suggestions/counter proposals and to hear about whatever you learn from LF about their past experiences.
I will try to join the office hours tomorrow to discuss the above and any next steps, mostly where to start the public discussion (governance repo? JEP?)

**Comment by krassowski** [comment]
> ![Image](https://github.com/user-attachments/assets/3ffad88a-1f71-4ec6-a5ba-31859e65de36)

Can we know what it is blocked on?

**Comment by Ruv7** [comment]
We are currently blocked on moving forward as of early March. The lawyer from LF and the lawyer from TIdelift team were in communication about the possibility of a new contract at that time but have not landed on a final answer or resolution. I have sent an email to ask for an update I can share.

**Comment by Carreau** [comment]
From a meeting a few weeks ago that I only attended for a few minutes due to other commitment I understood this was not a priority at the moment, and that the amounts were likely not worth the efforts.

My main concern is that AFAICT, there was no discussions between Jupyter-EC/LF and the rest of the community/maintainers to know for each project who would do the tidelift work – because there are regular work to actually do, and there might be a profound misunderstanding of what needs to be done/how things work.

The income provided by tidelift was several time referred to as "tidelift donation", while it is not in fact a donation. For example I currently lift [Flit](https://github.com/pypa/flit), and instead of having the money go  to my account, taxed, and then given to PSF, it happens to go directly to the PSF. Same for project I lifted for Jupyter for the past few years, it's not funds Tidelift donated to the project, it's funds _I_ donated to the project in exchange to work for tidelift.

I see no objections by going through LF, but I do no see how this can be solved without involving the people that will _actually do the work_ as they _need_ to have an account on tidelift, and if not this means the maintainers needs to actually have a process with the LF.


After all the complexity to add tidelift via LF that I curenlty see, I'm starting to retract my willingness to keep lifting the projects and send the money to LF if it's such a burden.

**Comment by Carreau** [comment]
I think we should close (I do not have the permissions to do so).

Following the acquisition of tidelift, https://tidelift.com/ is starting to redirect to https://www.sonarsource.com/solutions/security/, so my assumption is that in the future all the tidelift agreement will be somehow cancelled/overhauled.

**Comment by Carreau** [comment]
Can this be closed? AFAICT you can't lift new projects anymore.

**Comment by Carreau** [comment]
Can this be closed ? Please ?

**Comment by krassowski** [comment]
I think the people who can close this are @andrii-i or @jupyter-governance/executive-council

**Comment by choldgraf** [comment]
Sounds good, closing this as wontfix 👍
